logger/logger.go

```diff
@@ -140,52 +129,65 @@ func (l *Middleware) Handler(next http.Handler) http.Handler {
 	return http.HandlerFunc(fn)
 }
 
-func (l *Middleware) getBodyAndUser(r *http.Request) (body string, user string) {
-	ctx := r.Context()
-	if ctx == nil {
-		return "", ""
-	}
+var reMultWhtsp = regexp.MustCompile(`[\s\p{Zs}]{2,}`)
 
-	if l.inLogFlags(Body) {
-		if content, err := ioutil.ReadAll(r.Body); err == nil {
-			body = string(content)
-			r.Body = ioutil.NopCloser(bytes.NewReader(content))
+func (l *Middleware) getBody(r *http.Request) string {
+	if !l.logBody {
+		return ""
+	}
 
-			if len(body) > 0 {
-				body = strings.Replace(body, "\n", " ", -1)
-				body = reMultWhtsp.ReplaceAllString(body, " ")
-			}
+	reader, body, hasMore, err := peek(r.Body, int64(l.maxBodySize))
+	if err != nil {
+		return ""
+	}
 
-			if len(body) > l.maxBodySize {
-				body = body[:l.maxBodySize] + "..."
-			}
-		}
+	// "The Server will close the request body. The ServeHTTP Handler does not need to."
+	// https://golang.org/pkg/net/http/#Request
+	// So we can use ioutil.NopCloser() to make io.ReadCloser.
+	// Note that below assignment is not approved by the docs:
+	// "Except for reading the body, handlers should not modify the provided Request."
+	// https://golang.org/pkg/net/http/#Handler
+	r.Body = ioutil.NopCloser(reader)
+
+	if len(body) > 0 {
+		body = strings.Replace(body, "\n", " ", -1)
+		body = reMultWhtsp.ReplaceAllString(body, " ")
 	}
 
-	if l.inLogFlags(User) && l.userFn != nil {
-		u, err := l.userFn(r)
-		if err == nil && u != "" {
-			user = u
-		}
+	if hasMore {
+		body += "..."
 	}
 
-	return body, user
+	return body
 }
 
-func (l *Middleware) inLogFlags(f Flag) bool {
-	for _, flg := range l.flags {
-		if (flg == All && f != None) || flg == f {
-			return true
-		}
+// peek the first n bytes as string
+func peek(r io.Reader, n int64) (reader io.Reader, s string, hasMore bool, err error) {
```

Comment:
Is this needed? With io.LimiterReader probably can achive the same but easier.

Reply:
It depends on what we mean by "the same". If we need to peek some portion of the request body without placing restrictions on the body size then this function is needed. After all it is quite strange when logger middleware imposes restrictions on the request body size.

Reply:
well, I meant some TeeReader + LimiterReader combo to achieve safer stream reading. Actually, I'm not sure if this possible, just an idea. From another hand - I'm not fully convinced if this even a problem of logger, or smth we should control from the outside

Reply:
fyi: I have added [SizeLimit middleware](https://github.com/go-pkgz/rest/blob/master/sizelimit.go#L11) limiting body size